packages/@react-spectrum/s2/src/Tabs.tsx

```diff
@@ -314,50 +302,353 @@ function TabLine(props: TabLineProps) {
-  let {
-    density = 'regular',
-    isDisabled,
-    disabledKeys,
-    orientation = 'horizontal'
-  } = props;
-  let domRef = useDOMRef(ref);
+const icon = style({
+  display: 'block',
+  flexShrink: 0,
+  '--iconPrimary': {
+    type: 'fill',
+    value: 'currentColor'
+  }
+});
+
+export function Tab(props: TabProps) {
+  let {density, labelBehavior} = useContext(InternalTabsContext) ?? {};
 
   return (
-    <RACTabs
+    <RACTab
       {...props}
-      ref={domRef}
+      // @ts-ignore
+      originalProps={props}
       style={props.UNSAFE_style}
-      className={renderProps => (props.UNSAFE_className || '') + tabs({...renderProps}, props.styles)}>
-      <Provider
-        values={[
-          [TabsContext, {density, isDisabled, disabledKeys, orientation}]
-        ]}>
+      className={renderProps => (props.UNSAFE_className || '') + tab({...renderProps, density, labelBehavior}, props.styles)}>
+      {({
+          // @ts-ignore
+          isMenu
+        }) => {
+        if (isMenu) {
+          return props.children;
+        } else {
+          return (
+            <Provider
+              values={[
+                [TextContext, {
+                  styles:
+                    style({
+                      order: 1,
+                      display: {
+                        labelBehavior: {
+                          hide: 'none'
+                        }
+                      }
+                    })({labelBehavior})
+                }],
+                [IconContext, {
+                  render: centerBaseline({slot: 'icon', styles: style({order: 0})}),
+                  styles: icon
+                }]
+              ]}>
+              {typeof props.children === 'string' ? <Text>{props.children}</Text> : props.children}
+            </Provider>
+          );
+        }
+      }}
+    </RACTab>
+  );
+}
+
+const tabPanel = style({
+  display: 'flex',
+  marginTop: 4,
+  marginX: -4,
+  paddingX: 4,
+  color: 'gray-800',
+  flexGrow: 1,
+  flexShrink: 1,
+  flexBasis: '[0%]',
+  minHeight: 0,
+  minWidth: 0,
+  overflow: 'auto'
+}, getAllowedOverrides({height: true}));
+
+export function TabPanel(props: TabPanelProps) {
+  let {showTabs} = useContext(CollapseContext);
+  let {selectedKey} = useContext(InternalTabsContext);
+  if (showTabs) {
+    return (
+      <AriaTabPanel
+        {...props}
+        style={props.UNSAFE_style}
+        className={(props.UNSAFE_className ?? '') + tabPanel(null, props.styles)} />
+    );
+  }
+
+  if (props.id !== selectedKey) {
+    return null;
+  }
+
+  return <CollapsedTabPanel {...props} />;
+}
+
+function CollapsedTabPanel(props: TabPanelProps) {
+  let {UNSAFE_style, UNSAFE_className = '', ...otherProps} = props;
+  let {menuId} = useContext(CollapseContext);
+  let ref = useRef(null);
+  let tabIndex = useHasTabbableChild(ref) ? undefined : 0;
+
+  return (
+    <Group
+      {...otherProps}
+      ref={ref}
+      aria-labelledby={menuId}
```

Comment:
I'm only getting "Tab selector" as the accessible name on this element. Seems like it should get the full name from the tabs (e.g. "Monarchy and Republic")?